File: frontend/src/api/profiles.ts
import type {
  Delegation,
  DelegationScopeType,
  DelegationStatus,
  EmploymentEvent,
  EmploymentEventType,
  Position,
  PositionAssignmentType,
  Profile,
  ProfileFieldDefinition,
  ProfileFieldPermission,
  ProfilePosition,
  ReportingLine,
  ReportingLineType,
} from '@/types/api'
import { http } from './http'

export interface CreateProfilePayload {
  user_id: string
  employee_no: string
  real_name: string
  department_id: string
  job_title?: string | null
  phone?: string | null
  hire_date?: string | null
  custom_fields?: Record<string, unknown>
}

export interface UpdateProfilePayload {
  employee_no?: string | null
  real_name?: string | null
  department_id?: string | null
  job_title?: string | null
  phone?: string | null
  hire_date?: string | null
  custom_fields?: Record<string, unknown> | null
}

export interface CreatePositionPayload {
  code: string
  name: string
  level?: string | null
  extra_metadata?: Record<string, unknown>
  is_active?: boolean
}

export interface CreateProfilePositionPayload {
  position_id: string
  department_id: string
  assignment_type: PositionAssignmentType
  is_primary: boolean
  starts_at: string
  ends_at?: string | null
}

export interface CreateReportingLinePayload {
  manager_user_id: string
  department_id?: string | null
  line_type: ReportingLineType
  is_primary: boolean
  starts_at: string
  ends_at?: string | null
}

export interface CreateEmploymentEventPayload {
  event_type: EmploymentEventType
  effective_date: string
  title: string
  summary?: string | null
  payload?: Record<string, unknown>
}

export interface CreateDelegationPayload {
  delegator_user_id: string
  delegate_user_id: string
  scope_type: DelegationScopeType
  scope_department_id?: string | null
  scope_filters?: Record<string, unknown>
  starts_at: string
  ends_at: string
}

export interface UpdateDelegationPayload {
  status?: DelegationStatus
  starts_at?: string | null
  ends_at?: string | null
  scope_department_id?: string | null
  scope_filters?: Record<string, unknown> | null
}

export interface CreateProfileFieldDefinitionPayload {
  field_key: string
  label: string
  field_type: string
  storage_target: string
  is_sensitive?: boolean
  config?: Record<string, unknown>
  is_active?: boolean
}

export interface UpdateProfileFieldDefinitionPayload {
  label?: string
  field_type?: string
  storage_target?: string
  is_sensitive?: boolean
  config?: Record<string, unknown>
  is_active?: boolean
}

export interface CreateProfileFieldPermissionPayload {
  field_definition_id: string
  subject_type: string
  subject_value?: string | null
  can_view?: boolean
  can_edit?: boolean
  scope_filters?: Record<string, unknown>
  priority?: number
}

export interface UpdateProfileFieldPermissionPayload {
  subject_type?: string
  subject_value?: string | null
  can_view?: boolean
  can_edit?: boolean
  scope_filters?: Record<string, unknown>
  priority?: number
}

export async function listProfiles(): Promise<Profile[]> {
  const { data } = await http.get<Profile[]>('/profiles')
  return data
}

export async function getProfile(userId: string): Promise<Profile> {
  const { data } = await http.get<Profile>(`/profiles/${userId}`)
  return data
}

export async function createProfile(payload: CreateProfilePayload): Promise<Profile> {
  const { data } = await http.post<Profile>('/profiles', payload)
  return data
}

export async function updateProfile(userId: string, payload: UpdateProfilePayload): Promise<Profile> {
  const { data } = await http.patch<Profile>(`/profiles/${userId}`, payload)
  return data
}

export async function listPositions(): Promise<Position[]> {
  const { data } = await http.get<Position[]>('/positions')
  return data
}

export async function createPosition(payload: CreatePositionPayload): Promise<Position> {
  const { data } = await http.post<Position>('/positions', payload)
  return data
}

export async function createProfilePosition(
  userId: string,
  payload: CreateProfilePositionPayload,
): Promise<ProfilePosition> {
  const { data } = await http.post<ProfilePosition>(`/profiles/${userId}/positions`, payload)
  return data
}

export async function createProfileReportingLine(
  userId: string,
  payload: CreateReportingLinePayload,
): Promise<ReportingLine> {
  const { data } = await http.post<ReportingLine>(`/profiles/${userId}/reporting-lines`, payload)
  return data
}

export async function createProfileEvent(
  userId: string,
  payload: CreateEmploymentEventPayload,
): Promise<EmploymentEvent> {
  const { data } = await http.post<EmploymentEvent>(`/profiles/${userId}/events`, payload)
  return data
}

export async function createDelegation(payload: CreateDelegationPayload): Promise<Delegation> {
  const { data } = await http.post<Delegation>('/delegations', payload)
  return data
}

export async function updateDelegation(
  delegationId: string,
  payload: UpdateDelegationPayload,
): Promise<Delegation> {
  const { data } = await http.patch<Delegation>(`/delegations/${delegationId}`, payload)
  return data
}

export async function listProfileFieldDefinitions(): Promise<ProfileFieldDefinition[]> {
  const { data } = await http.get<ProfileFieldDefinition[]>('/profile-field-definitions')
  return data
}

export async function createProfileFieldDefinition(
  payload: CreateProfileFieldDefinitionPayload,
): Promise<ProfileFieldDefinition> {
  const { data } = await http.post<ProfileFieldDefinition>('/profile-field-definitions', payload)
  return data
}

export async function updateProfileFieldDefinition(
  definitionId: string,
  payload: UpdateProfileFieldDefinitionPayload,
): Promise<ProfileFieldDefinition> {
  const { data } = await http.patch<ProfileFieldDefinition>(
    `/profile-field-definitions/${definitionId}`,
    payload,
  )
  return data
}

export async function listProfileFieldPermissions(
  definitionId: string,
): Promise<ProfileFieldPermission[]> {
  const { data } = await http.get<ProfileFieldPermission[]>(
    `/profile-field-definitions/${definitionId}/permissions`,
  )
  return data
}

export async function createProfileFieldPermission(
  payload: CreateProfileFieldPermissionPayload,
): Promise<ProfileFieldPermission> {
  const { data } = await http.post<ProfileFieldPermission>('/profile-field-permissions', payload)
  return data
}

export async function updateProfileFieldPermission(
  permissionId: string,
  payload: UpdateProfileFieldPermissionPayload,
): Promise<ProfileFieldPermission> {
  const { data } = await http.patch<ProfileFieldPermission>(
    `/profile-field-permissions/${permissionId}`,
    payload,
  )
  return data
}

File: backend/app/schemas/profiles.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.core.enums import (
  DelegationScopeType,
  DelegationStatus,
  EmploymentEventTriggerStatus,
  EmploymentEventType,
  PositionAssignmentType,
  ReportingLineType,
  UserStatus,
)


class ProfileFieldAccessRead(BaseModel):
  field_key: str
  label: str
  field_type: str
  storage_target: str
  is_sensitive: bool
  value: Any
  can_view: bool
  can_edit: bool


class PositionRead(BaseModel):
  model_config = ConfigDict(from_attributes=True)

  id: UUID
  code: str
  name: str
  level: str | None
  extra_metadata: dict[str, Any]
  is_active: bool
  created_at: datetime
  updated_at: datetime


class ProfilePositionRead(BaseModel):
  model_config = ConfigDict(from_attributes=True)

  id: UUID
  user_id: UUID
  position_id: UUID
  department_id: UUID
  assignment_type: PositionAssignmentType
  is_primary: bool
  starts_at: date
  ends_at: date | None
  created_at: datetime
  updated_at: datetime


class ReportingLineRead(BaseModel):
  model_config = ConfigDict(from_attributes=True)

  id: UUID
  user_id: UUID
  manager_user_id: UUID
  department_id: UUID | None
  line_type: ReportingLineType
  is_primary: bool
  starts_at: date
  ends_at: date | None
  created_at: datetime
  updated_at: datetime


class EmploymentEventRead(BaseModel):
  model_config = ConfigDict(from_attributes=True)

  id: UUID
  user_id: UUID
  event_type: EmploymentEventType
  effective_date: date
  title: str
  summary: str | None
  payload: dict[str, Any]
  task_template_id: UUID | None
  workflow_definition_id: UUID | None
  trigger_status: EmploymentEventTriggerStatus
  triggered_at: datetime | None
  trigger_error: str | None
  trigger_attempt_count: int
  triggered_template_instance_id: UUID | None
  triggered_workflow_instance_id: UUID | None
  created_by: UUID
  created_at: datetime


class DelegationRead(BaseModel):
  model_config = ConfigDict(from_attributes=True)

  id: UUID
  delegator_user_id: UUID
  delegate_user_id: UUID
  scope_type: DelegationScopeType
  scope_department_id: UUID | None
  scope_filters: dict[str, Any]
  status: DelegationStatus
  starts_at: datetime
  ends_at: datetime
  created_by: UUID
  created_at: datetime
  updated_at: datetime


class ProfileFieldDefinitionRead(BaseModel):
  model_config = ConfigDict(from_attributes=True)

  id: UUID
  field_key: str
  label: str
  field_type: str
  storage_target: str
  is_sensitive: bool
  config: dict[str, Any]
  is_active: bool
  created_at: datetime
  updated_at: datetime


class ProfileFieldPermissionRead(BaseModel):
  model_config = ConfigDict(from_attributes=True)

  id: UUID
  field_definition_id: UUID
  subject_type: str
  subject_value: str | None
  can_view: bool
  can_edit: bool
  scope_filters: dict[str, Any]
  priority: int
  created_at: datetime
  updated_at: datetime


class ProfileRead(BaseModel):
  user_id: UUID
  user_email: str | None
  user_status: UserStatus | None
  employee_no: str | None
  real_name: str | None
  department_id: UUID | None
  job_title: str | None
  phone: str | None
  hire_date: date | None
  custom_fields: dict[str, Any]
  visible_fields: list[ProfileFieldAccessRead] = Field(default_factory=list)
  positions: list[ProfilePositionRead] = Field(default_factory=list)
  reporting_lines: list[ReportingLineRead] = Field(default_factory=list)
  employment_events: list[EmploymentEventRead] = Field(default_factory=list)
  delegations: list[DelegationRead] = Field(default_factory=list)
  created_at: datetime
  updated_at: datetime


class ProfileCreateRequest(BaseModel):
  user_id: UUID
  employee_no: str = Field(min_length=1, max_length=64)
  real_name: str = Field(min_length=1, max_length=120)
  department_id: UUID
  job_title: str | None = Field(default=None, max_length=120)
  phone: str | None = Field(default=None, max_length=32)
  hire_date: date | None = None
  custom_fields: dict[str, Any] = Field(default_factory=dict)


class ProfileUpdateRequest(BaseModel):
  employee_no: str | None = Field(default=None, min_length=1, max_length=64)
  real_name: str | None = Field(default=None, min_length=1, max_length=120)
  department_id: UUID | None = None
  job_title: str | None = Field(default=None, max_length=120)
  phone: str | None = Field(default=None, max_length=32)
  hire_date: date | None = None
  custom_fields: dict[str, Any] | None = None


class PositionCreateRequest(BaseModel):
  code: str = Field(min_length=1, max_length=64)
  name: str = Field(min_length=1, max_length=120)
  level: str | None = Field(default=None, max_length=64)
  extra_metadata: dict[str, Any] = Field(default_factory=dict)
  is_active: bool = True


class ProfilePositionCreateRequest(BaseModel):
  position_id: UUID
  department_id: UUID
  assignment_type: PositionAssignmentType = PositionAssignmentType.PRIMARY
  is_primary: bool = False
  starts_at: date
  ends_at: date | None = None


class ReportingLineCreateRequest(BaseModel):
  manager_user_id: UUID
  department_id: UUID | None = None
  line_type: ReportingLineType
  is_primary: bool = False
  starts_at: date
  ends_at: date | None = None


class EmploymentEventCreateRequest(BaseModel):
  event_type: EmploymentEventType
  effective_date: date
  title: str = Field(min_length=1, max_length=255)
  summary: str | None = None
  payload: dict[str, Any] = Field(default_factory=dict)
  task_template_id: UUID | None = None
  workflow_definition_id: UUID | None = None


class DelegationCreateRequest(BaseModel):
  delegator_user_id: UUID
  delegate_user_id: UUID
  scope_type: DelegationScopeType = DelegationScopeType.DATA_ACCESS
  scope_department_id: UUID | None = None
  scope_filters: dict[str, Any] = Field(default_factory=dict)
  starts_at: datetime
  ends_at: datetime


class DelegationUpdateRequest(BaseModel):
  status: DelegationStatus | None = None
  starts_at: datetime | None = None
  ends_at: datetime | None = None
  scope_department_id: UUID | None = None
  scope_filters: dict[str, Any] | None = None


class ProfileFieldDefinitionCreateRequest(BaseModel):
  field_key: str = Field(min_length=1, max_length=64)
  label: str = Field(min_length=1, max_length=120)
  field_type: str = Field(min_length=1, max_length=32)
  storage_target: str = Field(min_length=1, max_length=32)
  is_sensitive: bool = False
  config: dict[str, Any] = Field(default_factory=dict)
  is_active: bool = True


class ProfileFieldDefinitionUpdateRequest(BaseModel):
  label: str | None = Field(default=None, min_length=1, max_length=120)
  field_type: str | None = Field(default=None, min_length=1, max_length=32)
  storage_target: str | None = Field(default=None, min_length=1, max_length=32)
  is_sensitive: bool | None = None
  config: dict[str, Any] | None = None
  is_active: bool | None = None


class ProfileFieldPermissionCreateRequest(BaseModel):
  field_definition_id: UUID
  subject_type: str = Field(min_length=1, max_length=32)
  subject_value: str | None = Field(default=None, max_length=64)
  can_view: bool = False
  can_edit: bool = False
  scope_filters: dict[str, Any] = Field(default_factory=dict)
  priority: int = 100


class ProfileFieldPermissionUpdateRequest(BaseModel):
  subject_type: str | None = Field(default=None, min_length=1, max_length=32)
  subject_value: str | None = Field(default=None, max_length=64)
  can_view: bool | None = None
  can_edit: bool | None = None
  scope_filters: dict[str, Any] | None = None
  priority: int | None = None
